### PR TITLE
Add DAP Step back support

### PR DESCRIPTION
**Pull request by Anthony-Eid**
Release Notes:

- When debugging with a DAP that supports step back Zed will now show a step button button

### CODE REVIEW

File: crates/debugger_ui/src/debugger_panel_item.rs [hunk 1]
```diff
@@ -807,6 +819,17 @@ impl Render for DebugPanelItem {
                                     .disabled(thread_status != ThreadStatus::Stopped)
                                     .tooltip(move |cx| Tooltip::text("Step out", cx)),
                             )
+                            .when(capabilities.supports_step_back.unwrap_or(false), |this| {
```
Comment (RemcoSmitsDev) [inline crates/debugger_ui/src/debugger_panel_item.rs]: Can we move before the step over? I think that makes a bit more sense based on how the icons look

File: crates/project/src/dap_store.rs [hunk 2]
```diff
@@ -1077,6 +1078,37 @@ impl DapStore {
...
+        client_id: &DebugAdapterClientId,
+        thread_id: u64,
+        granularity: SteppingGranularity,
+        cx: &mut ModelContext<Self>,
+    ) -> Task<Result<()>> {
+        let Some(client) = self.client_by_id(client_id) else {
+            return Task::ready(Err(anyhow!("Could not find client: {:?}", client_id)));
+        };
+
+        let capabilities = self.capabilities_by_id(client_id);
```
Comment (RemcoSmitsDev) [inline crates/project/src/dap_store.rs:1092]: can you return and task::ready when the debug client does not support step back?

File: crates/debugger_ui/src/debugger_panel_item.rs (hunk 1 above)
Comment (Anthony-Eid) [reply crates/debugger_ui/src/debugger_panel_item.rs]: Good point! Will move it

File: crates/project/src/dap_store.rs (hunk 2 above)
Comment (Anthony-Eid) [reply crates/project/src/dap_store.rs:1092]: I’ll make the change. I didn’t before because the button isn’t even visible when it’s not supported

Comment (RemcoSmitsDev) [reply crates/project/src/dap_store.rs:1092]: Yeah agree, but the action is still there!

**Comment by RemcoSmitsDev** [comment]
Great work btw! But I think we also have to add a listener inside debugger_ui/lib so you can also use the action.

**Comment by Anthony-Eid** [comment]
@RemcoSmitsDev I don't think we should be setting the thread status before sending stepping requests on button clicks. If the request fails there's no way for a user to click on a button because they'll be disabled. We should set the thread status on receiving a positive response or reset the thread status when receiving a failed response  (I like the second idea).

This is something I ran into when testing the step back button for a DAP that doesn't support it, so I wanted to mention it here. It's scope is different than this PR so we should make a separate PR to address this issue.

**Comment by RemcoSmitsDev** [comment]
Yeah, that is something we have to think about, I like the idea of making it disabled, so the user cannot spam the buttons and that we give early feedback to the user about the action he did. So changing the status back when we receive an error from these actions would be a great solution to fix this.

**Pull request merged**